core/src/main/scala/chisel3/naming/Identifier.scala

```diff
@@ -0,0 +1,49 @@
+package chisel3.naming
+
+trait IdentifierProposer[T] {
+  def propose(obj: T): String
+}
```

Comment:
```suggestion
sealed trait IdentifierProposer[T] {
  def propose(obj: T): String
}
```

Since it doesn't seem to be useful for users to extend this, I think it should be private. Some day I hope we can make it user implementable, but I don't want to give the appearance of doing so when we don't yet.